### PR TITLE
Reservation page

### LINKED ISSUE #7: Create Reservation page [2pts]
Should include the following:
- Cancel all active appointments
- Active appointment:
  - Should include an Cancel btn.
  - Button to edit date (Re-Schedule) (Optional)
- History:
  - Past appointments (Optional)
  - Canceled appointments (Optional)

### PR DESCRIPTION
**Pull request by richardoppiyo**
In this milestone, I worked on the following:
 - Cancel all active appointments
 - Active appointment:
    -Should include an Cancel btn.
 - Button to edit date (Re-Schedule) (Optional)
 - History:
    -Past appointments (Optional)
   - Canceled appointments (Optional)

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -13,8 +14,9 @@ const App = () => (
     <Routes>
       <Route path="/" element={<Layout />}>
         <Route index element={<Home />} />
+        <Route path="/lawyers/:id" element={<Lawyer />} />
+        <Route path="/reservations" element={<Reservations />} />
```
Comment (helio3197) [inline src/App.js]: Currently, the routes for the `Lawyer `and `Reservations` are children of the route with path "/", so for all of their child routes, prepend the "/" in the path prop is not necessary.
```suggestion
        <Route path="lawyers/:id" element={<Lawyer />} />
        <Route path="reservations" element={<Reservations />} />
```

File: src/redux/reservations/deleteReservation.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
+const API_RESERVATIONS_INDEX_ENDPOINT = `${process.env.REACT_APP_API_HOST}/api/v1/reservations`;
+const REQUEST_STARTED = 'book-a-lawyer/reservations/REQUEST_STARTED';
+const REQUEST_FAILED = 'book-a-lawyer/reservations/REQUEST_FAILED';
+const REQUEST_COMPLETED = 'book-a-lawyer/reservations/REQUEST_COMPLETED';
+const initialState = {
+  reservations: [],
+};
```
Comment (helio3197) [inline src/redux/reservations/deleteReservation.js]: This part is exactly the same as the `reservation.js` reducer, the action types are the same, which may cause unexpected behaviors. A good idea is renaming the action types.
Also a good idea is having an initial status for the state.
```suggestion
const API_RESERVATIONS_DELETE_ENDPOINT = `${process.env.REACT_APP_API_HOST}/api/v1/reservations`;
const REQUEST_STARTED = 'book-a-lawyer/reservationDelete/REQUEST_STARTED';
const REQUEST_FAILED = 'book-a-lawyer/reservationDelete/REQUEST_FAILED';
const REQUEST_COMPLETED = 'book-a-lawyer/reservationDelete/REQUEST_COMPLETED';
const initialState = {
  reservations: [],
  status: 'idle',
};
```

File: src/redux/reservations/deleteReservation.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  },
+});
+
+const requestCompleted = ({ reservations }) => ({
+  type: REQUEST_COMPLETED,
+  payload: {
+    status: 'completed',
+    reservations,
+  },
+});
```
Comment (helio3197) [inline src/redux/reservations/deleteReservation.js]: The delete reservation action doesn't return an object, so having a `reservation` parameter is unnecessary.
```suggestion
const requestCompleted = () => ({
  type: REQUEST_COMPLETED,
  payload: {
    status: 'completed',
  },
});
```

File: src/redux/reservations/deleteReservation.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+export const deleteReservations = (id) => async (dispatch, getState) => {
+  dispatch(requestStarted());
+  try {
+    const { authToken } = getState().auth;
+    const response = await fetch(`${API_RESERVATIONS_INDEX_ENDPOINT}/${id}`, {
```
Comment (helio3197) [inline src/redux/reservations/deleteReservation.js]: Update the API endpoint.
```suggestion
    const response = await fetch(`${API_RESERVATIONS_DELETE_ENDPOINT}/${id}`, {
```

**Review by helio3197 — CHANGES_REQUESTED** [review]
Hi @richardoppiyo :wave: you have made an outstanding job so far congrats! :+1:

**To highlight:** :checkered_flag: 

- No linter errors :heavy_check_mark:
- Professional PR :heavy_check_mark: 
- Clean code :heavy_check_mark:

There are some **required changes** to make your App work better.

**Required changes:**
- Check the comments below

I don't have anything more to tell you other than wish you good luck and happy coding!

**Comment by richardoppiyo** [comment]
Thank you for the recommendations. I have worked on them.

File: src/redux/reservations/reservations.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    reservations,
+  },
+});
+
+export const clearFetchReservationState = () => ({
+  type: CLEAR_FETCH_STATE,
+  payload: {
+    status: 'idle',
+  },
+});
```
Comment (helio3197) [inline src/redux/reservations/reservations.js:67]: In the specific case of fetching the reservation_index, we want to keep the list of fetched reservations and it should persist on the state through the session lifetime, having an action for clearing the state in this reducer may cause unnecessary re-fetching the reservations resource every time a component using this reducer is mounted.

```suggestion
```

File: src/redux/reservations/reservations.js [hunk 6]
```diff
@@ -0,0 +1,84 @@
+const API_RESERVATIONS_INDEX_ENDPOINT = `${process.env.REACT_APP_API_HOST}/api/v1/reservations`;
+const REQUEST_STARTED = 'book-a-lawyer/reservations/REQUEST_STARTED';
+const REQUEST_FAILED = 'book-a-lawyer/reservations/REQUEST_FAILED';
+const REQUEST_COMPLETED = 'book-a-lawyer/reservations/REQUEST_COMPLETED';
+const CLEAR_FETCH_STATE = 'book-a-lawyer/reservationsDelete/CLEAR_FETCH_STATE';
```
Comment (helio3197) [inline src/redux/reservations/reservations.js:5]: As in the comment above having a clear state action on this reducer may be unnecessary.

```suggestion
```

File: src/components/Reservations/Reservations.js [hunk 7]
```diff
@@ -0,0 +1,131 @@
+/* eslint-disable max-len */
+import React, { useEffect, useState } from 'react';
+import Button from 'react-bootstrap/Button';
+import { useDispatch, useSelector } from 'react-redux';
+import { useSearchParams } from 'react-router-dom';
+import Container from 'react-bootstrap/Container';
+import Spinner from 'react-bootstrap/Spinner';
+import Pagination from 'react-bootstrap/Pagination';
+import { fechReservations } from '../../redux/reservations/reservations';
+import { deleteReservations, clearDeleteReservationState, clearFetchReservationState } from '../../redux/reservations/deleteReservation';
```
Comment (helio3197) [inline src/components/Reservations/Reservations.js]: Importing `clearFetchReservationState` may not be necessary here. Please kindly remove it. 

```suggestion
import { deleteReservations, clearDeleteReservationState } from '../../redux/reservations/deleteReservation';
```

File: src/components/Reservations/Reservations.js [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+  useEffect(() => {
+    if (reservationState.status !== 'completed') {
+      dispatch(fechReservations());
+    }
+    if (lawyersState.status !== 'completed') {
+      dispatch(getLawyers());
+    }
+    return () => {
+      dispatch(clearFetchReservationState());
```
Comment (helio3197) [inline src/components/Reservations/Reservations.js]: After the component is unmounted, the action for reseting the state should be dispatched.

```suggestion
      dispatch(clearDeleteReservationState());
```

File: src/components/Reservations/Reservations.js [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+    if (maxPagesNumber && (showCurrentPage > maxPagesNumber)) {
+      setSearchParams({ p: maxPagesNumber });
+      setShowCurrentPage(maxPagesNumber);
+    }
+  }, [maxPagesNumber]);
+
+  const deleteReservationData = (id) => () => {
+    dispatch(deleteReservations(id));
+    dispatch(clearDeleteReservationState());
+  };
```
Comment (helio3197) [inline src/components/Reservations/Reservations.js:50]: Here you are dispatching both actions at the same time, which may cause unexpected behaviors or errors. Please kindly remove that behaviour.
```suggestion
  const deleteReservationData = (id) => () => {
    dispatch(deleteReservations(id));
  };
```

**Review by helio3197 — CHANGES_REQUESTED** [review]
Hi @richardoppiyo you have made a good job so far! 👍 

There are some changes to be implemented, please check the comments below.

**Review by helio3197 — APPROVED** [review]
Thank you for applying the requested changes 👍

**Pull request merged**